src/checkers/inference/solver/util/PrintUtils.java

```diff
@@ -159,146 +151,212 @@ private static void buildStatisticText(String key, Integer value, StringBuilder
+     * Write the statistics to a file called statistics.txt.
+     *
+     * @param statistics a map between stats keys and their long values.
+     * @param modelRecord
+     * @param appendWrite if set to true, will append the solutions to the output file.
+     */
+    public static void writeStatistics(Map<StatisticKey, Long> statistics,
+            Map<String, Integer> modelRecord,
+            boolean appendWrite) {
+        File outFile = new File("statistics.txt");
+        PrintStream out = getFilePrintStream(outFile, appendWrite);
+        outputStatistics(out, statistics, modelRecord);
+        out.close();
+        System.out.println("Statistics have been written to: " + outFile.getAbsolutePath() + "\n");
+    }
+
+
+    private static String generateUnsolveableString(Collection<Constraint> unsatisfactoryConstraints) {
         ToStringSerializer toStringSerializer = new ToStringSerializer(false);
         SlotPrinter slotPrinter = new SlotPrinter(toStringSerializer);
+
+        StringBuffer sb = new StringBuffer();
+
         // Print constraints and related slots
-        System.out.println("\n=================================== Explanation Starts=================================\n");
-        System.out.println("------------------ Unsatisfactory Constraints ------------------\n");
+        sb.append("------------------ Unsatisfactory Constraints ------------------\n");
         for (Constraint constraint : unsatisfactoryConstraints) {
-            System.out.println("\t" + constraint.serialize(toStringSerializer) + " \n\t    " + constraint.getLocation().toString() + "\n");
+            sb.append("\t" + constraint.serialize(toStringSerializer)
+                    + " \n\t\t" + constraint.getLocation().toString() + "\n");
         }
-        System.out.println("------------- Related Slots -------------\n");
-        for (Constraint c : unsatisfactoryConstraints) {
-            c.serialize(slotPrinter);
+        sb.append("------------- Related Slots -------------\n");
+        for (Constraint constraint : unsatisfactoryConstraints) {
+            sb.append(constraint.serialize(slotPrinter));
+        }
+
+        return sb.toString();
+    }
+
+    /**
+     * Outputs unsatisfactory constraints to the given stream.
+     * @param stream
+     * @param unsatisfactoryConstraints
+     */
+    private static void outputUnsolveable(PrintStream stream, Collection<Constraint> unsatisfactoryConstraints) {
+        stream.println("/***********************Explanation************************/");
+        stream.println(generateUnsolveableString(unsatisfactoryConstraints));
+        stream.println("/**********************************************************/");
+    }
+
+    /**
+     * Print the unsolveable constraints and their related slots to screen.
+     */
+    public static void printUnsolvable(Collection<Constraint> unsatisfactoryConstraints) {
+        if (unsatisfactoryConstraints == null || unsatisfactoryConstraints.isEmpty()) {
```

Comment:
this should be moved to call site, let util method be just about the printing / writing